core/server/worker/src/main/java/alluxio/worker/grpc/AbstractReadHandler.java

```diff
@@ -126,11 +153,46 @@ public void onNext(alluxio.grpc.ReadRequest request) {
       mContext.setPosReceived(mContext.getRequest().getStart());
       mDataReaderExecutor.submit(createDataReader(mContext, mResponseObserver));
       mContext.setDataReaderActive(true);
+    } catch (RejectedExecutionException e) {
+      handleStreamEndingException(e, Status.RESOURCE_EXHAUSTED);
     } catch (Exception e) {
-      LogUtils.warnWithException(LOG, "Exception occurred while processing read request {}.",
-          request, e);
-      mSerializingExecutor.execute(() -> mResponseObserver
-          .onError(AlluxioStatusException.fromCheckedException(e).toGrpcStatusException()));
+      handleStreamEndingException(
+          AlluxioStatusException.fromThrowable(e).toGrpcStatusException());
+    }
+  }
+
+  private void handleStreamEndingException(StatusException e) {
+    handleStreamEndingException(e, e.getStatus());
+  }
+
+  /**
+   * Handle any exception which should abort the client's read request.
+   *
+   * @param exception the exception thrown
+   * @param status the type of {@link Status} exception which should be returned to the user
+   */
+  private void handleStreamEndingException(Exception exception, Status status) {
+    Long sessionId = mContext.getRequest() == null ? null : mContext.getRequest().getSessionId();
+    LogUtils.warnWithException(LOG, "fatal error occurred while handling read. sessionId: {}",
+        sessionId, exception);
+    AlluxioStatusException statusExc = new AlluxioStatusException(status);
+    try (LockResource lr = new LockResource(mLock)) {
+      if (mContext == null) {
+        mContext = createRequestContext(alluxio.grpc.ReadRequest.newBuilder().build());
+      }
+      setError(new Error(statusExc, true));
+    }
+    // After calling setError, this will run the data reader to complete the request and close
+    // any possibly opened blocks, then replies to the client with onError, and should eventually
+    // shut down the serializing executor.
+    // If possible, we should free up this thread immediately by running the data reader in a
+    // separate thread because we want to avoid the possibility of blocking the `onNext` method
```

Comment:
We shouldn't care.
Thread here is allocated from gRPC's internal no-limit pool, just to serve onNext() and terminate afterward.

Reply:
I was told previously that the "no limit" pool was merely for simple RPCs. 

For the stream handling aspect, the number of threads designated to serve the stream observers are coming from the workers netty threadpool, which is finite.

Reply:
I'd expect that only under `directExecutor()` mode which is not the case here.
Netty is supposed to do proto parsing and allocate a thread from the gRPC server executor for calling `StreamObserver<>` callbacks with deserialized prototypes.

Reply:
@bf8086 and @calvinjia I think have said something otherwise, or maybe I misunderstood. Maybe we should all have a short discussion?